tests/unittests/test_runs/test_fallback_datasource.py

```diff
@@ -0,0 +1,240 @@
+        self.remove_seed_dir(seed_dir)
+        self.reRoot(self.tmp)
+        initer = stages.Init(ds_deps=['FILESYSTEM'])
+        iid_without_dsrc = self.cloud_init(initer)
+        self.assertEqual(iid_with_dsrc, iid_without_dsrc)
+
+    def test_invalid_openstack_uri(self):
+        _register_uris(self.VERSION, {}, {}, OS_FILES)
+        sys_cfg = {
+            'datasource_list': ['OpenStack'],
+            'datasource': {'OpenStack': {'dsmode': 'net'}}
+        }
+        cloud_cfg = safeyaml.dumps(sys_cfg)
+        util.ensure_dir(os.path.join(self.tmp, 'etc', 'cloud'))
+        util.write_file(os.path.join(self.tmp, 'etc',
+                                     'cloud', 'cloud.cfg'), cloud_cfg)
+        self.ds = dsos.DataSourceOpenStack
+        ds_os = self.ds(sys_cfg=sys_cfg, distro=None, paths=self.paths)
+        MOCK_PATH = 'cloudinit.sources.DataSourceOpenStack.'
+        mock_path = MOCK_PATH + 'detect_openstack'
+        with tests_helpers.mock.patch(mock_path) as m_detect_os:
+            m_detect_os.return_value = True
+            found = ds_os.get_data()
+        self.assertTrue(found)
+        self.reRoot(self.tmp)
+        # simulate first boot with datasource openstack
+        initer = stages.Init()
+        initer.datasource = ds_os
+        iid_with_dsrc = self.cloud_init(initer)
+
+        # simulate subsequent boot without datasource openstack
+        util.del_file(os.path.join(self.tmp, "run/cloud-init/.instance-id"))
+        hp.register_uri(hp.GET, re.compile(r'http://169.254.169.254/.*'),
+                        status=504)
+        self.reRoot(self.tmp)
+        initer = stages.Init()
+        iid_without_dsrc = self.cloud_init(initer)
+        self.assertEqual(iid_with_dsrc, iid_without_dsrc)
+
+    def test_invalid_ovf(self):
+        self.ds = dsovf.DataSourceOVF
+        props = {"password": "passw0rd", "instance-id": "inst-001"}
+        env = fill_properties(props)
+        ovf_env = self.tmp_path('ovf-env.xml', dir=self.paths.seed_dir)
+        util.write_file(ovf_env, env)
+        dsrc = self.ds(sys_cfg={}, distro=None, paths=self.paths)
+        MPATH = 'cloudinit.sources.DataSourceOVF.'
+        with mock.patch(MPATH + 'transport_vmware_guestinfo') as m_guestd:
+            with mock.patch(MPATH + 'transport_iso9660') as m_iso9660:
+                m_iso9660.return_value = None
+                m_guestd.return_value = None
+                self.assertTrue(dsrc.get_data())
+
+        self.reRoot(self.tmp)
+        # simulate first boot with datasource OVF
+        initer = stages.Init()
+        initer.datasource = dsrc
+        iid_with_dsrc = self.cloud_init(initer)
+        # simulate subsequent boot without datasource OVF
+        self.remove_seed_dir(self.paths.seed_dir)
+        self.reRoot(self.tmp)
+        initer = stages.Init(ds_deps=['FILESYSTEM'])
+        iid_without_dsrc = self.cloud_init(initer)
+        self.assertEqual(iid_with_dsrc, iid_without_dsrc)
```

Comment:
Looking at tall test_invalid_ cases; it appears to be the same test; I suspect that you could have a common test function, the key parts:

1) mock datasource such that it ds.get_data() returns True
2) do first boot and return instance-id
3) mock datasource such that ds.get_data() returns False forcing FallbackDS path
4) do second boot and return instance-id
5) assert iid from (2) == iid from (4)

Reply:
I tried to reuse code as much as possible with latest commit. Let me know if this can be further improvised if I am missing something.